File: java/vertx/lib/src/main/java/reis/polyglot/lib/apiResponses/ErrorResponse.java
package reis.polyglot.lib.apiResponses;

import io.vertx.core.json.JsonObject;
import reis.polyglot.lib.exceptions.ApplicationException;
import reis.polyglot.lib.exceptions.ApplicationExceptionFactory;

public class ErrorResponse extends ApiResponse {
    private final ApplicationException error;

    private ErrorResponse(ApplicationException error) {
        super(error.getStatusCode(), false);
        this.error = error;
    }

    public static ErrorResponse fromException(ApplicationException exception) {
        return new ErrorResponse(exception);
    }

    public static ErrorResponse notFound() {
        return ErrorResponse.fromException(ApplicationExceptionFactory.NotFound());
    }

    @Override
    public JsonObject toJson() {
        return super.toJson()
            .putNull("data")
            .put("error", this.error.toJson());
    }
}

File: servers/typescript/express/src/api/v1/lib/responses/error-response.ts
import { Response } from 'express';
import { ApplicationError, ApplicationException } from '@lib/exceptions/application-exception';
import { ApiResponse } from './api-response';

export class ErrorResponse extends ApiResponse {
    constructor(private readonly error: ApplicationError) {
        super(error.statusCode, false);
    }

    override send(res: Response) {
        const error = (this.error instanceof ApplicationException) ? this.error.toJson() : this.error;
        this.formatAndSend(res, error, null);
    }
}
